rust/src/normalization.rs

```diff
@@ -120,6 +121,294 @@ pub(crate) fn rewrite_rpmdb_timestamps<F: Read + Write + Seek>(rpmdb: &mut F) ->
+                Exec::cmd("db_dump")
+                    .args(&[PROC_SELF_CWD.join(&path)])
+                    .cwd(PROC_SELF_FD.join(rootfs.as_raw_fd().to_string()))
+                    .stdout(Redirection::Pipe)
+                    .capture()
+                    .context("pre-normalization data dump")?
+                    .stdout
+                    .as_slice(),
+            );
+
+            {
+                let mut file_id = sha2::Sha256::new();
+                file_id.update(timestamp.to_be_bytes());
+                file_id.update(format!("rpm/{}", entry.file_name().to_str().unwrap()).as_bytes());
+                let file_id = &file_id.finalize()[..20];
+
+                let mut db = rootfs.update_file(&path, 0o644)?;
+                let meta_header: MetaHeader = db.read_le()?;
+
+                match (meta_header.magic, meta_header.page_type) {
+                    (BTREE_MAGIC, PageType::BTreeMeta) => (),
+                    (HASH_MAGIC, PageType::HashMeta) => (),
+                    _ => continue,
+                };
+
+                db.seek(SeekFrom::Start(PAGE_HEADER_MAGIC_OFFSET))?;
+                db.write_all(file_id)?;
+
+                for pageno in 1..meta_header.last_pgno + 1 {
+                    db.seek(SeekFrom::Start((pageno * meta_header.pagesize) as u64))?;
+
+                    let header: PageHeader = db.read_le()?;
+
+                    if header.page_type == PageType::Overflow {
+                        db.seek(SeekFrom::Current(header.hf_offset as i64))?;
+                        let fill_length =
+                            meta_header.pagesize - (PAGE_HEADER_SIZE + header.hf_offset) as u32;
+                        std::io::copy(
+                            &mut std::io::repeat(b'\x00').take(fill_length as u64),
+                            &mut db,
+                        )?;
+                        continue;
+                    }
+
+                    let mut offsets: Vec<u16> = Vec::new();
+
+                    for _ in 0..header.entries {
+                        offsets.push(db.read_le()?);
+                    }
+
+                    offsets.sort_unstable();
+
+                    let empty = if offsets.is_empty() {
+                        meta_header.pagesize - PAGE_HEADER_SIZE as u32
+                    } else {
+                        *offsets.first().unwrap() as u32
+                            - (PAGE_HEADER_SIZE + header.entries * 2) as u32
+                    };
+
+                    std::io::copy(&mut std::io::repeat(b'\x00').take(empty as u64), &mut db)?;
+
+                    let mut offset_iter = offsets.into_iter().peekable();
+                    while let Some(offset) = offset_iter.next() {
+                        db.seek(SeekFrom::Start(
+                            (pageno * meta_header.pagesize + offset as u32) as u64,
+                        ))?;
+
+                        if matches!(header.page_type, PageType::IBTree | PageType::LBTree) {
```

Comment:
Honestly this is pretty impressive.  I am curious again - did you just read through the BDB docs and write this from scratch, with some trial and error?  Was there any inspiration from related code in bdb?

Basically any comments you can think of for someone 2 years from now who has to fix a bug in this and needs some background information would certainly be appreciated :smile:

Reply:
It started with hexdumps and wondering what the hell was making things vary, and then realising that a bunch of stuff was random stuff that I couldn't work out where it was coming from, and then reading the source code and cursing everyone involved.